# Урок 8

# Сингулярное разложение матриц

In [1]:
import numpy as np
np.set_printoptions(precision=2, suppress=True)

## Практическое задание

1. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$


2. Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    
    б) норму Фробениуса.

In [2]:
A = np.array([[1, 2, 0],
              [0, 0, 5],
              [3, -4, 2],
              [1, 6, 5],
              [0, 1, 0]])
print(f'Матрица A:\n{A}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [3]:
U, s, W = np.linalg.svd(A)

In [4]:
print(f'Матрица U:\n{U}')

Матрица U:
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]


In [5]:
print(f'Матрица s:\n{s}')

Матрица s:
[8.82 6.14 2.53]


In [6]:
print(f'Матрица W:\n{W}')

Матрица W:
[[ 0.07  0.72  0.69]
 [-0.37  0.67 -0.65]
 [-0.93 -0.21  0.31]]


In [7]:
print(np.dot(U.T, U))

[[ 1.  0. -0.  0. -0.]
 [ 0.  1.  0.  0.  0.]
 [-0.  0.  1. -0. -0.]
 [ 0.  0. -0.  1. -0.]
 [-0.  0. -0. -0.  1.]]


In [8]:
V = W.T

D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [9]:
print(f'Матрица D:\n{D}')

Матрица D:
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]


- Евклидова норма $$\left \| A \right \|_{E}=\mu_{1} = 8.82$$

- норма Фробениуса $$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}.$$

In [10]:
norm_frob = sum(num**2 for num in s)** 0.5

In [11]:
print(f'норма Фробениуса A = {norm_frob}')

норма Фробениуса A = 11.045361017187263
